lib/docassemble/docassemble_api_interface.js

```diff
@@ -0,0 +1,377 @@
+};  // Ends da_i.try_api_key()
+
+
+// node -e 'require("./lib/github_integration/docassemble_api_interface.js").loop_to_create_project()'
+da_i.loop_to_create_project = async function( create_project_options ) {
+  /** Create a docassemble Project on the docassemble server's account.
+  *    Loop through potential project names until we find one that
+  *    isn't already taken and then use that name to create the Project. */
+  log.info({ pre: `Trying to create a docassemble project.` });
+
+  let default_opts = { timeout: 3 * 1000, max_tries: 15, wait_seconds: 5 };
+  let { timeout, max_tries, wait_seconds } = create_project_options || default_opts;
+
+  let project_name = null;
+  let name_incrementor = 0;
+  let response;
+
+  // Try to create a project. If the project name is already taken, increment and loop again.
+  while ( name_incrementor <= max_tries ) {
+    name_incrementor++;
+    try {
+
+      // Try next name that could be unique. Must remove all non-valid characters
+      project_name = session_vars.get_project_base_name().replace(/[^A-Za-z0-9]+/, '') + name_incrementor;
+      log.info({ pre: `Trying to create a Project with the name ${ project_name }` });
+      response = await _da.create_project( project_name, timeout );
+      log.info({ pre: `The docassemble server was ready. Created a Project named ${ project_name }` });
+      // If succeeded in creating project, return the name that worked
+      return project_name;
+
+    } catch ( error ) {
+      response = error.response;
+
+      if ( error.response && error.response.data ) {
+        if ( error.response.data.includes( `Invalid project` ) ) {
+          // A non-unique project name should just cause another loop
+          continue;
+
+        } else {
+          // Other response errors should be thrown with link to docs
+          throw `${ error.response.status }: ${ error.response.data } See https://docassemble.org/docs/api.html#playground_post_project for more details.`;
+        }
+
+      } else {
+
+        // If it's a timeout error, server is busy. Try again after waiting.
+        if ( is_timeout_error( error ) && name_incrementor < max_tries ) {
+          log.info({ pre: `The docassemble server was busy when trying to create project. `
+            + `Attempt ${ name_incrementor } of ${ max_tries }. Will try again `
+            + `after ${ wait_seconds } seconds.` });
+          await time.waitForTimeout( wait_seconds * 1000 );
+          continue;
+
+        // A non-timeout, non-server error should just be thrown
+        } else {
+          log.warn({ pre: `Ran into unexpected error when trying to create project.` });
+          throw error;
+
+        }  // ends type of non-response error
+      }  // ends type of reponse error
+    }  // ends try
+  }  // ends while
+
+  // Failed because there were too many project names already taken
+  if ( response.data && response.data.includes( `Invalid project` ) && name_incrementor > max_tries) {
```

Comment:
Project, not project.

Reply:
Ah. Actually, this is not a message to the user. This is actually the text that docassemble sends back in the error and we do have to check for that exact text unless we use regex with the `i` flag.

Reply:
Sorry that I missed that when we were discussing it.